File: MyHashMap.java
import java.util.HashSet;
import java.util.LinkedList;
import java.util.Map;
import java.util.Set;

public class MyHashMap<KEY, VALUE> {
   public static final int STARTING_MAX_SIZE = 4;
   public static final double LOAD_FACTOR = 0.75;

   private int currentMaxSize;
   private int currentSize;
   private LinkedList<Pair<KEY, VALUE>>[] list;


   public MyHashMap() {
      currentMaxSize = STARTING_MAX_SIZE;
      currentSize = 0;
      list = (LinkedList<Pair<KEY, VALUE>>[]) new LinkedList[STARTING_MAX_SIZE];
   }

   public void put(KEY newKey, VALUE newVal) {
       int hashKey = hash(newKey);
       LinkedList<Pair<KEY, VALUE>> old = list[hashKey];

       boolean keyAlreadyExists = false;
       int indexOfKey = -1;

       if (old != null) {
          for (int i = 0; i < old.size(); i++) {
             KEY keyVal = old.get(i).getKey();
             if (keyVal.equals(newKey)) {
                keyAlreadyExists = true;
                indexOfKey = i;
                break;
             }
          }
       }

       if (keyAlreadyExists) {
          old.set(indexOfKey, new Pair<>(newKey, newVal));
       } else {
          list[hashKey] = new LinkedList<Pair<KEY, VALUE>>();
          list[hashKey].add(new Pair<>(newKey, newVal));
          currentSize++;
       }
       double fillRatio = currentSize / (double) currentMaxSize; // how full the hashmap is
       if (fillRatio > LOAD_FACTOR) {
          rehashAll();
       }
   }

   public VALUE get(KEY key) {
      int listIndex = -1;
      int hashKey = hash(key);
      LinkedList<Pair<KEY, VALUE>> itemsList = list[hashKey];
      for (Pair<KEY, VALUE> pair : itemsList) {
         KEY currentKey = pair.getKey();
         if (currentKey.equals(key)) {
            return pair.getValue();
         }
      }
      return null;
   }

   public void remove(KEY key) {
      int hashKey = hash(key);
      LinkedList<Pair<KEY, VALUE>> sublist = list[hashKey];
      for (int i = 0; i < sublist.size(); i++) {
         Pair<KEY, VALUE> pair = sublist.get(i);
         KEY currentKey = pair.getKey();
         if (currentKey.equals(key)) {
            sublist.remove(i);
            return;
         }
      }
   }

   public Set<KEY> keySet() {
      Set<KEY> keySet = new HashSet<>();
      for (LinkedList<Pair<KEY, VALUE>> sublist : list) {
         if (sublist == null) {
            continue;
         }
         for (Pair<KEY, VALUE> pair : sublist) {
            keySet.add(pair.getKey());
         }
      }
      return keySet;
   }

   public String toString() {

      StringBuilder string = new StringBuilder();
      int listLength = list.length;
      System.out.println("list length: " + listLength);
      for (int i = 0; i < list.length; i++) {
         System.out.println("i: " + i);
         LinkedList<Pair<KEY, VALUE>> sublist = list[i];

         if (sublist == null) {
            continue;
         }
         int sublistSize = sublist.size();
         for (int j = 0; j < sublistSize; j++) {
            System.out.println("j: " + j);
            Pair<KEY, VALUE> pair = sublist.get(j);

            string.append(pair.getKey().toString()).append(" -> ")
                  .append(pair.getValue().toString()).append("\n");
         }
      }
      int lastIndex = string.length();
      string.delete(lastIndex-2, lastIndex);
      return string.toString();
   }

   public void debugPrint() {
      for (LinkedList<Pair<KEY, VALUE>> sublist : list) {
         System.out.println(sublist);
      }
   }

   private void rehashAll() {
      currentMaxSize *= 2;
      LinkedList<Pair<KEY, VALUE>>[] newList = (LinkedList<Pair<KEY, VALUE>>[]) new LinkedList[currentMaxSize];
      for (LinkedList<Pair<KEY, VALUE>> sublist : list) {
         if (sublist == null) {
            continue;
         }
         for (Pair<KEY, VALUE> pair : sublist) {
            KEY key = pair.getKey();
            int newHash = hash(key);
            newList[newHash] = new LinkedList<Pair<KEY, VALUE>>();
            newList[newHash].add(pair);
         }
      }
      list = newList;
   }

   private int hash(KEY key) {
      return key.hashCode() % currentMaxSize;
   }


}

File: python/my_hashmap.py
from typing import TypeVar
from collections.abc import Hashable


def main():
    hashmap = MyHashMap()
    hashmap["Hello"] = "World"
    hashmap["Chase"] = "IS awesome"
    hashmap[1] = [1, 2, 3]
    hashmap["test"] = "testing"
    hashmap["hi"] = "idi"
    hashmap["test"] = "new val for test"
    del hashmap["Chase"]
    item = hashmap["test"]
    print(hashmap, end="\n\n")
    #for key in hashmap:
        #print(key)
    for key, value in hashmap.items():
        print(f"key: {key} | value: {value}")



KEY = TypeVar("KEY", bound=Hashable)
VALUE = TypeVar("VALUE")

STARTING_MAX_SIZE = 2
LOAD_FACTOR = 0.75


class MyHashMap:
    def __init__(self):
        self._current = 0  # only used for iterating
        self._list: list[list[tuple[KEY, VALUE]]] = [[] for _ in range(STARTING_MAX_SIZE)]
        self._current_max_size: int = STARTING_MAX_SIZE
        self._current_size: int = 0

    def _put(self, new_key, new_val) -> None:
        hash_key: int = self._hash(new_key)
        old: list[tuple[KEY, KEY]] = self._list[hash_key]

        key_already_exists = False
        index_of_key = -1
        for index, (key, val) in enumerate(old):
            if key == new_key:
                key_already_exists = True
                index_of_key = index
                break

        if key_already_exists:
            index: int = index_of_key
            self._list[hash_key][index] = (new_key, new_val)
        else:
            self._list[hash_key].append((new_key, new_val))
        if not old:  # if old is empty
            self._current_size += 1
        if (self._current_size / self._current_max_size) > LOAD_FACTOR:
            self._rehash_all()

    def _get(self, key) -> VALUE:
        list_index = None

        hash_key: int = self._hash(key)
        items_list: list[tuple[KEY, VALUE]] = self._list[hash_key]
        for og_key, val in items_list:
            if og_key == key:
                return val
        raise KeyError(f"key {key} not found")

    def _remove(self, key) -> None:
        hash_key: int = self._hash(key)
        sublist: list[tuple[KEY, VALUE]] = self._list[hash_key]
        for index, (og_key, val) in enumerate(sublist):
            if key == og_key:
                sublist.pop(index)
                return
        raise KeyError(f"Key {key} not in HashMap")

    def _rehash_all(self) -> None:
        self._current_max_size *= 2
        new_list: list[list[tuple[KEY, VALUE]]] = [[] for _ in range(self._current_max_size)]
        for value_list in self._list:  # 5439543 -> [("hello", "word"), ("chase", "is cool")]
            if not value_list:
                continue
            for key_val_tuple in value_list:
                key, val = key_val_tuple
                new_hash: int = self._hash(key)
                new_list[new_hash].append(key_val_tuple)
        self._list = new_list

    def _hash(self, val) -> int:
        return hash(val) % self._current_max_size

    def _key_set(self) -> list[KEY]:
        key_set: list["KEY"] = []
        for key_list in self._list:
            if not key_list:
                continue
            for key, val in key_list:
                key_set.append(key)
        return key_set

    def _debug_print(self) -> None:
        string = ""
        for index, sublist in enumerate(self._list):
            if not sublist:
                continue
            string += f"hash: {index} | {sublist}\n"
        print(string)

    def __str__(self):
        string = ""
        for sublist in self._list:
            if sublist == [[]]:
                continue
            for og_key, value in sublist:
                string += f"{og_key} -> {value}\n"
        return string

    def __setitem__(self, key: KEY, value: VALUE):
        return self._put(key, value)

    def __getitem__(self, key: KEY) -> VALUE:
        return self._get(key)

    def __iter__(self) -> KEY:
        for key in self._key_set():
            yield key

    def items(self) -> tuple[KEY, VALUE]:
        for key_list in self._list:
            if not key_list:
                continue
            for key_val_pair in key_list:
                yield key_val_pair

    def __delitem__(self, key: KEY) -> None:
        self._remove(key)


if __name__ == "__main__":
    main()
